apps/emqx_durable_storage/src/emqx_ds_cache.hrl

```diff
@@ -0,0 +1,36 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+-ifndef(EMQX_DS_CACHE_HRL).
+-define(EMQX_DS_CACHE_HRL, true).
+
+-define(CACHE_KEY(LASTKEY), {LASTKEY}).
+-define(EOS_KEY, []).
+-define(EOS_MARKER, []).
+
+-record(cache_entry, {
+    key :: ?CACHE_KEY(emqx_ds:message_key()) | ?EOS_KEY,
+    seqno :: seqno(),
+    inserted_at :: timestamp(),
```

Comment:
Can we use the timestamp from the key? We don't have to be very precise in the cache eviction flow.

Reply:
From the key inside the message, you mean?  If we assume only more or less recent keys are added to cache (i.e., we don't start caching from the distant past), I guess so.

If it's the `#cache_entry.key` you mean, I think there's no way to extract it from the `emqx_ds:message_key()` in the general case without yet another callback.  :sweat_smile: